packages/core/src/mixins/primitives/reatomEnum.ts

```diff
@@ -0,0 +1,82 @@
+import {
+  action,
+  Action,
+  Atom,
+  atom,
+  ReatomError,
+} from 'src/core'
+
+export type EnumFormat = 'camelCase' | 'snake_case'
+
+export type EnumAtom<
+  T extends string,
+  Format extends EnumFormat = 'camelCase',
+> = Atom<T> & {
+  [Variant in T as Format extends 'camelCase'
+    ? `set${Capitalize<Variant>}`
+    : Format extends 'snake_case'
+      ? `set_${Variant}`
+      : never]: Action<[], Variant>
+} & {
+  reset: Action<[], T>
+  enum: { [K in T]: K }
+}
+
+export type EnumAtomOptions<
+  T extends string,
+  Format extends EnumFormat = 'camelCase',
+> = {
+  name?: string
+  format?: Format
+  initState?: T
+}
+
+export const reatomEnum = <
+  const T extends string,
+  Format extends 'camelCase' | 'snake_case' = 'camelCase',
+>(
+  variants: ReadonlyArray<T>,
+  options: string | EnumAtomOptions<T, Format> = {},
+) => {
+  const {
+    name,
+    format = 'camelCase' as Format,
+    initState = variants[0],
+  }: EnumAtomOptions<T, Format> = typeof options === 'string'
+    ? { name: options }
+    : options
+
+  const stateAtom = atom(initState, name) as EnumAtom<T, Format>
+  const enumAtom: typeof stateAtom = Object.assign((...args: any[]) => {
```

Comment:
This version of Reatom DO NOT ACCEPT (hightly not recomended) to change or create by yourself the atom reference, **use middleware instead**.